### PR TITLE
Porting adaptive receive buffer prediction

### PR DESCRIPTION
**Pull request by nayato**
Motivation:
Provides good flexible default while lowering memory consumption on average.

Modifications:
- Synced IRecvByteBufAllocator and relevant parts up to date
- Ported AdaptiveRecvByteBufAllocator
- Extra: properly porting ChannelMetadata

Result:
Process of receiving bytes from network is more configurable and allows for both more conservative and more agressive ingestion.

### CODE REVIEW

**Comment by nayato** [comment]
@myagley, @mtuchkov PTAL

File: src/DotNetty.Transport/Channels/DefaultMaxMessagesRecvByteBufAllocator.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    ///     and also prevents overflow.
+    /// </summary>
+    public abstract class DefaultMaxMessagesRecvByteBufAllocator : IMaxMessagesRecvByteBufAllocator
+    {
+        volatile int maxMessagesPerRead;
+
+        protected DefaultMaxMessagesRecvByteBufAllocator()
+            : this(1)
+        {
+            ;
```
Comment (myagley) [inline src/DotNetty.Transport/Channels/DefaultMaxMessagesRecvByteBufAllocator.cs]: Is this semicolon intended?

File: src/DotNetty.Transport/Channels/DefaultMaxMessagesRecvByteBufAllocator.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+        protected DefaultMaxMessagesRecvByteBufAllocator(int maxMessagesPerRead)
+        {
+            this.MaxMessagesPerRead = maxMessagesPerRead;
+        }
+
+        public int MaxMessagesPerRead
+        {
+            get { return this.maxMessagesPerRead; }
+            set { this.maxMessagesPerRead = value; }
```
Comment (myagley) [inline src/DotNetty.Transport/Channels/DefaultMaxMessagesRecvByteBufAllocator.cs]: Should this be checked for negatives?

File: src/DotNetty.Transport/Channels/DefaultMaxMessagesRecvByteBufAllocator.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    && this.AttemptedBytesRead == this.lastBytesRead
+                    && this.totalMessages < this.maxMessagePerRead
+                    && this.totalBytesRead < int.MaxValue;
+            }
+
+            public virtual void ReadComplete()
+            {
+            }
+
+            public int AttemptedBytesRead { get; set; }
```
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/DefaultMaxMessagesRecvByteBufAllocator.cs]: virtual?

File: src/DotNetty.Transport/Channels/DefaultMaxMessagesRecvByteBufAllocator.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                    // The value may be "invalid" after this point, but it doesn't matter because reading will be stopped.
+                    this.totalBytesRead += value;
+                    if (this.totalBytesRead < 0)
+                    {
+                        this.totalBytesRead = int.MaxValue;
+                    }
+                }
+            }
+
+            public bool ContinueReading()
```
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/DefaultMaxMessagesRecvByteBufAllocator.cs]: virtual?

**Comment by mtuchkov** [comment]
LPITM (Looks pretty identical to me) :) :+1:

**Comment by myagley** [comment]
LGTM except for the two comments I had.

**Pull request merged**